src/rpc/rawtransaction.cpp

```diff
@@ -1613,45 +1464,45 @@ UniValue signrawtransaction(const UniValue& params, bool fHelp)
 
     vector<unsigned char> txData(ParseHexV(params[0], "argument 1"));
     CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
+    CDataStream ssVersion(txData, SER_NETWORK, PROTOCOL_VERSION);
     vector<CMutableTransaction> txVariants;
-    while (!ssData.empty()) {
-        try {
-            CMutableTransaction tx;
-            ssData >> tx;
-            txVariants.push_back(tx);
-        }
-        catch (const std::exception&) {
-            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
-        }
-    }
+    CMutableScCertificate certificate;
 
-    if (txVariants.empty())
-        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transaction");
+    if (ssData.empty())
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing input transaction or certificate");
 
-    // mergedTx will end up with all the signatures; it
-    // starts as a clone of the rawtx:
-    CMutableTransaction mergedTx(txVariants[0]);
+    int txVersion;
+    ssVersion >> txVersion;
 
-    // Fetch previous transactions (inputs):
-    CCoinsView viewDummy;
-    CCoinsViewCache view(&viewDummy);
-    {
-        LOCK(mempool.cs);
-        CCoinsViewCache &viewChain = *pcoinsTip;
-        CCoinsViewMemPool viewMempool(&viewChain, mempool);
-        view.SetBackend(viewMempool); // temporarily switch cache backend to db+mempool view
-
-        BOOST_FOREACH(const CTxIn& txin, mergedTx.vin) {
-            const uint256& prevHash = txin.prevout.hash;
-            CCoins coins;
-            view.AccessCoins(prevHash); // this is certainly allowed to fail
+    if(txVersion != SC_CERT_VERSION) {
+        while (!ssData.empty()) {
+            try {
+                CMutableTransaction tx;
+                ssData >> tx;
+                txVariants.push_back(tx);
+            }
+            catch (const std::exception& ex) {
+                throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction decode failed");
+            }
         }
+    } else {
+        try {
+            ssData >> certificate;
 
-        view.SetBackend(viewDummy); // switch back to avoid locking mempool for too long
+            if (!ssData.empty()) {
+                // just one and only one certificate expected
+                throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("Found %d extra byte%safter certificate",
+                ssData.size(), ssData.size()>1?"s ":" "));
+            }
+        }
+        catch (const std::exception& ex) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction and certificate decode failed");
```

Comment:
Change message to "Certificate decode failed" because we are expect the Cert here.